content/en/hosting/4.x/data-migration-k8s.md

```diff
@@ -0,0 +1,282 @@
+---
+title: "Migration from CHT 3.x to CHT 4.x in Kubernetes"
+linkTitle: "K8s Data Migration to 4.x"
+weight: 1
+aliases:
+  - /apps/guides/hosting/4.x/k8s-data-migration
+description: >
+  Guide to migrate existing data from CHT 3.x to CHT 4.x in Kubernetes environments
+relatedContent: >
+---
+
+The hosting architecture differs entirely between CHT-Core 3.x and CHT-Core 4.x. When both versions are running in Kubernetes, migrating data requires specific steps using the `couchdb-migration` tool. This tool interfaces with CouchDB to update shard maps and database metadata.
+
+{{% alert title="Note" %}}
+If after upgrading you get an error, `Cannot convert undefined or null to object` - please see [issue #8040](https://github.com/medic/cht-core/issues/8040) for a work around. This only affects CHT 4.0.0, 4.0.1, 4.1.0 and 4.1.1. It was fixed in CHT 4.2.0.
+{{% /alert %}}
+
+## Migration Steps
+
+### 1. Initial Setup
+```shell
+# Set your namespace
+NAMESPACE=<your-namespace>
+```
+
+### 2. Access the Source CouchDB (3.x)
+```shell
+# List all pods to find medic-os pod
+kubectl get pods -n $NAMESPACE
+
+# Access the medic-os pod
+kubectl exec -it -n $NAMESPACE <medic-os-pod-name> -- bash
+```
+
+### 3. Set Up Migration Tool Inside 3.x Pod
+
+Once inside the pod, install required dependencies:
+```shell
+# Install Node.js
+curl -fsSL https://deb.nodesource.com/setup_16.x | bash -
+apt install -y nodejs
+
+# Install npm (if not installed with Node.js)
+apt install -y npm
+
+# Install git and clone repository
+apt update
+apt install -y git
+git clone https://github.com/medic/couchdb-migration.git
+cd couchdb-migration
+npm ci --omit=dev
+```
+
+### 4. Run Pre-Migration Commands on 3.x
+
+Get credentials from 1Password and set them inside the pod:
+```shell
+export ADMIN_USER=<admin_username_from_1password>
+export ADMIN_PASSWORD=<admin_password_from_1password>
+
+# Set COUCH_URL
+export COUCH_URL="http://${ADMIN_USER}:${ADMIN_PASSWORD}@localhost:5984"
+
+# Verify connection
+curl -s $COUCH_URL/_up
+```
+
+Pre-index views to minimize downtime:
+```shell
+./bin/pre-index-views <desired CHT version>
+```
+
+{{% alert title="Note" %}} 
+If pre-indexing is omitted, 4.x API will fail to respond to requests until all views are indexed. For large databases, this could take many hours or days.
+{{% /alert %}}
+
+Save CouchDB configuration:
+```shell
+./bin/get-env
+```
+
+Save the output containing:
+- CouchDB secret (used for encrypting passwords and session tokens)
+- CouchDB server UUID (used for replication checkpointing)
+- CouchDB admin credentials
+
+### 5. Clone the 3.x Data Volume
+
+First, identify the volume ID from your 3.x data:
+```shell
+# Get the PVC name
+kubectl get pvc -n $NAMESPACE
+
+# Get the volume ID from the PVC
+VOLUME_ID=$(kubectl get pvc <your-3x-pvc-name> -n $NAMESPACE -o jsonpath='{.spec.volumeName}')
+EBS_VOLUME_ID=$(kubectl get pv $VOLUME_ID -o jsonpath='{.spec.awsElasticBlockStore.volumeID}' | cut -d'/' -f4)
```

Comment:
I think because this section is so long, it's likely having them manually set  `<your-3x-pvc-name>` to be an env var.  let's  break this out into it's own step to find the the PVC name and set the three env vars. the next larger section can then just by copy and pasted blindly